design/cluster-api-provider-metal3/multi-tenancy_contract.md

```diff
@@ -0,0 +1,289 @@
+namespace as the Metal3Cluster, containing a kubeconfig file. The optional
+`context` should point to a context in the kubeconfig file. If the `context` is
+not specified, the current context will be used.
+
+For the Metal3MachineTemplate, the new field will be
+`.spec.template.spec.identityRef` with the same sub-fields as the Metal3Cluster.
+This will propagate to the Metal3Machine's `.spec.identityRef`.
+
+If the `identityRef` is not specified, CAPM3 should use in-cluster credentials
+for accessing the BareMetalHosts (current behavior). If the `identityRef` is
+specified on the Metal3Cluster, but not on the Metal3MachineTemplate, CAPM3
+should use the credentials specified in the Metal3Cluster for all
+Metal3Machines.
+
+CAPM3 should make use of the kubeconfig file to access the BareMetalHosts
+related to the Metal3Cluster. The kubeconfig file can point to a namespace in
+the same cluster or to a remote cluster. Secrets for metadata, networkdata and
+userdata must also be created next to the BareMetalHosts.
+
+Example:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3Cluster
+metadata:
+  name: cluster-1
+spec:
+  identityRef:
+    name: cluster-1-kubeconfig
+    context: cluster-1-context
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3MachineTemplate
+metadata:
+  name: cluster-1-controlplane-template
+spec:
+  template:
+    spec:
+      identityRef:
+        name: control-plane-bmh-kubeconfig
+        context: default
+```
+
+The status of the Metal3Machine should be extended with the information from the
+BareMetalHost. This is similar to how other providers work, as they propagate
+the relevant information from cloud resources back the the InfraMachine. For the
+Metal3Machine, we should add the following fields from the BareMetalHost:
+
+- `.status.provisioning` on the BMH would become
+  `.status.bareMetalHost.provisioning` (ID is omitted)
+- `.status.poweredOn` on the BMH would become `.status.bareMetalHost.poweredOn`
+- `.status.errorMessage` on the BMH would become
+  `.status.bareMetalHost.errorMessage`
+- `.status.operationalStatus` on the BMH would become
+  `.status.bareMetalHost.operationalStatus`
+
+Fields related to the BMC credentials are not included, since they are only
+relevant for managing the BareMetalHost. Hardware details are also avoided since
+they are about to be removed from the BareMetalHost status now that we have a
+separate CRD (HardwareData) for them. We could think about propagating the
+information from the HardwareData or even copy the whole thing if deemed useful.
+
+The status of the Metal3Cluster does not need to change. However, the Ready
+condition should be set only when CAPM3 has verified that the identityRef
+credentials are valid. Invalid credentials should result in an error on the
+Metal3Cluster.
+
+Example Metal3Machine status:
+
+```yaml
+status:
+  # New fields
+  bareMetalHost:
```

Comment:
I assume this example is outdated per discussion above?

Reply:
Yes I need to update this

Reply:
Updated!